### ISSUE TITLE
NaNs seen with transformers version >= 4.21.0 when running HF BERT fine-tuning with XLA_USE_BF16=1

### ISSUE THREAD

**Issue by jeffhataws** [pytorch/xla#4152]
## 🐛 Bug

When running HuggingFace BERT (any size) fine-tuning tutorial with transformers version >= 4.21.0 and using XLA_USE_BF16=1, I see NaNs in the loss after the first step. The issue is seen also with XLA_DOWNCAST_BF16=1.

## To Reproduce

Use latest nightly matching your platform:
```
docker pull gcr.io/tpu-pytorch/xla:r1.13_3.7_cuda_11.2
docker run -it --shm-size 16G --gpus all gcr.io/tpu-pytorch/xla:r1.13_3.7_cuda_11.2 
```

Use a tiny BERT and only for 1 step for faster run:
```
pip install -U transformers==4.24.0 datasets sklearn
cd ~/
git clone https://github.com/huggingface/transformers --branch v4.24.0
cd transformers/examples/pytorch/text-classification/

tee run > /dev/null << EOF
export TASK_NAME=mrpc
GPU_NUM_DEVICES=1 XLA_USE_BF16=1 python3 ./run_glue.py \
--model_name_or_path prajjwal1/bert-tiny \
--task_name $TASK_NAME \
--do_train \
--max_seq_length 128 \
--per_device_train_batch_size 8 \
--learning_rate 2e-5 \
--max_steps 1 \
--overwrite_output_dir \
--output_dir /tmp/$TASK_NAME/ |& tee log_run
EOF

source ./run
```

```
***** train metrics *****                                                                                                                                    
  epoch                    =        0.0                                                                                                                      
  train_loss               =        nan                                                                                                                      
  train_runtime            = 0:00:03.09                                                                                                                      
  train_samples            =       3668                                                                                                                      
  train_samples_per_second =      2.585                                                                                                                      
  train_steps_per_second   =      0.323                                                                                                                      
```

## Expected behavior

Expected results (with transformers version 4.20.0):
```
***** train metrics *****
  epoch                    =        0.0
  train_loss               =     0.7031
  train_runtime            = 0:00:03.08
  train_samples            =       3668
  train_samples_per_second =      2.595
  train_steps_per_second   =      0.324
```
## Environment

 - Reproducible on XLA backend [CPU/TPU]: GPU
 - torch_xla version: 1.13

## Additional context

The NaNs likely come from the transformers library change: https://github.com/huggingface/transformers/pull/17306 . This PR replaced many lines which used to be -float(inf) (or other small constants) with torch.finfo(<some data type>).min. For torch.float32 the min value is -3.4028234663852886e+38 which is smaller than the bfloat16 minimum of -3.3895313892515355e+38. So we need to inspect where in torch/xla the conversion to BF16 could introduce NaN or Inf (since Inf could lead to NaN via IEEE rule Inf * 0.0 = NaN).

I have examined the graphs produced in version 4.24.0 compared to 4.20.0 (before the change), and see that the graphs are identical, meaning that the NaN or Inf is introduced at the inputs.

**Comment by jeffhataws** [comment]
Here's a simpler test, and only involve inference/forward pass:

```
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification, AutoConfig
import transformers
import os
import warnings

tokenizer = AutoTokenizer.from_pretrained("prajjwal1/bert-tiny")
model = AutoModelForSequenceClassification.from_pretrained("prajjwal1/bert-tiny", return_dict=False)
# model.to(torch.bfloat16)
model.to("xla")

sequence_0 = "The company HuggingFace is based in New York City"
sequence_2 = "HuggingFace's headquarters are situated in Manhattan"

max_length=128
paraphrase = tokenizer.encode_plus(sequence_0, sequence_2, max_length=max_length, padding='max_length', truncation=True, return_tensors="pt")
paraphrase.to("xla")

input_ids = paraphrase['input_ids']
attention_mask = paraphrase['attention_mask']
token_type_ids = paraphrase['token_type_ids']
example_inputs_paraphrase = input_ids, attention_mask, token_type_ids

# print(example_inputs_paraphrase)
paraphrase_classification_logits_neuron = model(*example_inputs_paraphrase)
print(paraphrase_classification_logits_neuron)
```

Run like this to dump HLO also:

```
XLA_SAVE_TENSORS_FILE=./dump XLA_SAVE_TENSORS_FMT=hlo XLA_DOWNCAST_BF16=1 GPU_NUM_DEVICES=1 python test.py
```
```
(tensor([[nan, nan]], device='xla:1', grad_fn=<AddmmBackward0>),)
```
Looking closely at the change https://github.com/huggingface/transformers/pull/17306 , if we experiment by modifying site-packages/transformers/modeling_utils.py to undo the change, like this:
```
        #extended_attention_mask = (1.0 - extended_attention_mask) * torch.finfo(dtype).min
        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
```
then we don't see NaNs anymore:
```
(tensor([[ 0.0547, -0.0742]], device='xla:1', grad_fn=<AddmmBackward0>),)
```

So the problem is that torch.finfo(torch.float32).min = -3.4028234663852886e+38 gets converted to -inf in torch/xla before the graph. When the original encoder_extended_attention_mask is 1, then encoder_extended_attention_mask becomes (1.0 - 1.0 ) * -inf which becomes NaN.

One potential solution is to add `model.to(torch.bfloat16)` in the above test code which would ensure torch.finfo(torch.bfloat16).min = -3.3895313892515355e+38 and not -inf . Then the results would not have Nans.

Another solution is to use HF's bf16 option. However, currently that option only supports CUDA, as you get the follow error "Your setup doesn't support bf16/gpu. You need torch>=1.10, using Ampere GPU with cuda>=11.0".

Since https://github.com/huggingface/transformers/pull/17306 is a fix for multiple bugs, it won't be reverted. However, we can find a solution that works for bf16.

**Comment by JackCaoG** [comment]
Thanks for the info, I will try to take a look this week or find a owner.

**Comment by JackCaoG** [comment]
@Liyang90 can you take a look? From the comment above I don't know if there is an easy way to fix it from pytorch/xla infra part(since it involves model code change, and nan was passed to pytorch/xla directly). Would this issue be more suitable for HF?

**Comment by Liyang90** [comment]
> @Liyang90 can you take a look? From the comment above I don't know if there is an easy way to fix it from pytorch/xla infra part(since it involves model code change, and nan was passed to pytorch/xla directly). Would this issue be more suitable for HF?

Yes, I agree. The issue is that the model code passes a value that is out of range for `XLA_USE_BF16=1`, so the conversion would cast it to `-inf`. Even in PyTorch itself, this is the expected behavior:
```
>>> a = torch.tensor(torch.finfo(torch.float32).min, dtype=torch.float32)
>>> a
tensor(-3.4028e+38)
>>> a.dtype
torch.float32
>>> b = a.to(torch.bfloat16)
>>> b
tensor(-inf, dtype=torch.bfloat16)
```
There is no special treatment to match `torch.finfo(torch.float32).min` with `torch.finfo(torch.bfloat16).min`.

The best thing that can be done here is check for `XLA_USE_BF16` env var and set `model.to(torch.bfloat16)` accordingly ahead of time.

**Comment by jeffhataws** [comment]
Thanks @Liyang90  @JackCaoG . One correct way is to overload ".to()" to check bounds and clip to the reduced precision bounds. However, this is expensive and not performant. Will focus on HF modifications instead.
